irodsConnector/manager.py

```diff
@@ -36,11 +34,21 @@ class IrodsConnector:
     _session = None
     _tickets = None
     _users = None
-    _ibridges_configuration = None
-    _irods_env_file = ''
-    _irods_environment = None
     use_icommands = None
 
+    def __new__(cls):
```

Comment:
Do we really have to make the manager a singleton?
In test application I am already instantiating up to 10 managers to speed up data transfers.

Reply:
I thought of multiple managers, but a single manager is capable of multiple connections and we can allow _it_ to manage those threads.  If we try to do it, we will inevitably need to add thread safety, something the PRC already does.  I just don't think the current arrangement is ideal for that.

Also, I have been testing alternate clients for our users, and like the PRC, all connections go through the connected catalog server, even if there are separate resource servers.  For those cases, this adds significant load that can be avoided with a single manager.

Eventually, the PRC will add forwarding to the resource server for optimal, multi-threaded transfers.  I think it better to wait for that than to add complexity to iBridges in this way.

In any case, how much improvement are you seeing?  I wonder if it can be repeated with multiple connections from a single manager.  Certainly worth a look?

Reply:
> Eventually, the PRC will add forwarding to the resource server for optimal, multi-threaded transfers.

PRC 'resource redirect' PR is here... https://github.com/irods/python-irodsclient/pull/458

eyeballs welcome!

Reply:
> Do we really have to make the manager a singleton?

Also, as discussed offline, when a singleton `IrodsConnector` is used, there is only ever one `irods.Session` object with its pool of connections, that _should_ be usable in a multi-threading environment.

@d-w-moore Is it safe to use in this way?

The alternative to a singleton connector would be _either_ to allow each class that needs its functionality to instantiate a _new_ `IrodsConnector` _or_ to pass the one canonical connector around via arguments as was the original case.

For the multi-connector scenario, there could be up to a dozen or more connectors, each one having to make a _new_ connection to the iRODS server, each with its associated authentication overhead when called upon.  And no telling what kind of thread collisions might happen if we are not careful enough.

As for the single connector scenario, it feels like taking a huge step backwards having to pass it around between classes and methods and functions.  The whole reason we chose to use a singleton Context object with embedded connector in the first place was to avoid exactly this requirement.